### ISSUE TITLE
matmul / linear: passing core_grid or program_config silently drops bf16 math fidelity from HiFi2 to LoFi

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#55889]
**Component / Area:** `ttnn/cpp/ttnn/operations/matmul/device/matmul_device_operation.cpp` (`create_matmul_attributes`); `ttnn.matmul`, `ttnn.linear`, `ttnn.addmm`, `ttnn.matmul_batched_weights`

**Issue Type:** Bug (silent numerics change)

## Observed

`ttnn.matmul(a, b)` on bfloat16 inputs runs at `MathFidelity.HiFi2`. The same call with a `core_grid=` or a `program_config=` runs at `LoFi`, and nothing in the docstrings, the docs or the returned tensor says so. The op reports it itself: `ttnn.create_matmul_attributes` resolves `HiFi2` with nothing given and `LoFi` with a program config given, on Blackhole and on Wormhole.

With the block config ttnn picks for `matmul(a, b)` on Blackhole passed back as an explicit `program_config`, the output changes in **27,899,151 of 29,360,128 words (95.0 %)** and the call gets 14 % faster; putting the fidelity back with `compute_kernel_config=HiFi2` makes it bit-identical to `matmul(a, b)` again, at the same speed. A user who passes a config or a grid to speed a matmul up is therefore measuring the fidelity, not the config, and a user who passes one for any other reason (sharding, a fused activation, a specific grid) loses one fidelity level without being told. A downstream fusion tracer of ours carried this for weeks as an unexplained ~15 k-word difference between `matmul` plus a fused chain and the eager ops.

Who reaches it: counting `ttnn.matmul(...)` / `ttnn.linear(...)` calls in `models/` by AST (shipped code, `tests/` excluded), 146 calls pass `core_grid` (98) or `program_config` (48) with no `compute_kernel_config` and so run at LoFi on bf16 inputs (e.g. `models/demos/bert/tt/ttnn_optimized_bert.py:21`, `models/demos/bert/tt/ttnn_optimized_sharded_bert.py:107`); 325 pass neither and run at HiFi2; 278 set `compute_kernel_config` and are unaffected either way.

The rule is one line in `create_matmul_attributes` (`matmul_device_operation.cpp:2683`): `increase_fidelity = !has_program_config && !has_user_grid && !are_inputs_low_precision_df`, HiFi2 when true and LoFi otherwise, with float32 inputs then overridden to HiFi4 (HiFi3 on Wormhole).

## Measured

`repro_matmul_fidelity.py` (below): 2048×4096×14336 and 32×4096×14336 bf16, random inputs scaled by `k^-0.5`, seed 17; 20 timed calls after 3 warm-up with `synchronize_device` around the loop; bit patterns compared, never PCC. `cfg` is the `MatmulMultiCoreReuseMultiCastProgramConfig` ttnn resolves for `matmul(a, b)` on the 11×10 Blackhole grid (per-core 7×41, `in0_block_w` 1, subblock 7×1), passed back explicitly. On Wormhole ttnn's own no-config choice is a different block config, so the Wormhole columns compare the explicit-config arms with each other only.

### Explicit program config, 2048×4096×14336: only `compute_kernel_config` moves

| `compute_kernel_config` | fidelity resolved | P300 us | P300 vs HiFi2 | N300 us | N300 vs HiFi2 | words differing from the HiFi2 arm | share of 29,360,128 |
|---|---|---|---|---|---|---|---|
| not given | LoFi | 1222.7 | -14.4% | 5889.0 | -12.2% | 27,899,151 | 95.0% |
| HiFi2 | HiFi2 | 1427.6 | 0.0% | 6707.7 | 0.0% | 0 | 0.0% |
| LoFi | LoFi | 1215.4 | -14.9% | 5888.6 | -12.2% | 27,899,151 | 95.0% |
| HiFi4 | HiFi4 | 2427.4 | +70.0% | 9179.8 | +36.9% | 23,569,350 | 80.3% |

The word count is the same on both architectures because `in0_block_w = 1` gives the same per-element accumulation order.

### Explicit config against `matmul(a, b)`, P300, where `cfg` is ttnn's own choice

| call | P300 us | vs `matmul(a, b)` | words differing from `matmul(a, b)` | share of 29,360,128 |
|---|---|---|---|---|
| `matmul(a, b)` | 1428.6 | 0.0% | 0 | 0.0% |
| `matmul(a, b, program_config=cfg)` | 1222.7 | -14.4% | 27,899,151 | 95.0% |
| `matmul(a, b, program_config=cfg, compute_kernel_config=HiFi2)` | 1427.6 | -0.1% | 0 | 0.0% |

### `core_grid` = the full compute grid, 32×4096×14336

| call | P300 us | P300 vs HiFi2 arm | N300 us | N300 vs HiFi2 arm | words differing from the HiFi2 arm | share of 458,752 |
|---|---|---|---|---|---|---|
| `matmul(a, b, core_grid=G)` | 293.9 | -0.3% | 628.1 | +2.2% | 436,516 | 95.2% |
| `matmul(a, b, core_grid=G, compute_kernel_config=HiFi2)` | 294.9 | 0.0% | 614.5 | 0.0% | 0 | 0.0% |
| `matmul(a, b)` | 294.9 | 0.0% | 614.9 | +0.1% | 0 | 0.0% |

### `core_grid` at 2048×4096×14336, N300: the speed-up a user sees is the fidelity

| call | N300 us | vs `matmul(a, b)` |
|---|---|---|
| `matmul(a, b)` | 5338.4 | 0.0% |
| `matmul(a, b, core_grid=G)` | 4612.9 | -13.6% |
| `matmul(a, b, core_grid=G, compute_kernel_config=HiFi2)` | 5525.4 | +3.5% |

On P300 the same `core_grid=G` call is 2950.5 us, 2.07× `matmul(a, b)`: the full-grid config ttnn builds there is not the default's block config and is slower at either fidelity (3074.3 us at HiFi2), so the fidelity is not separable by bits at that shape, and the 32-row shape above carries the `core_grid` claim.

Ruled out: the block config (first table, the same `cfg` in every row); `packer_l1_acc` / `fp32_dest_acc_en` (the HiFi2 arm that restores bit-exactness uses the defaults, `packer_l1_acc=True`, `fp32_dest_acc_en=False`); bfp8_b / bfp4_b inputs (LoFi in every form, unaffected).

<details>
<summary><code>repro_matmul_fidelity.py</code> (torch + ttnn only; <code>TT_DEV=&lt;n&gt; python repro_matmul_fidelity.py</code>)</summary>

```python
# ttnn.matmul(a, b) on bf16 inputs runs at MathFidelity::HiFi2; the same call with
# core_grid= or program_config= runs at LoFi (create_matmul_attributes). This script
#   1. asks create_matmul_attributes which fidelity it resolves with and without a program config,
#   2. times each call form and counts output words that differ from matmul(a, b).
#
# TT_DEV=<n> python repro_matmul_fidelity.py
import os
import time

import torch
import ttnn

device = ttnn.open_device(device_id=int(os.environ.get("TT_DEV", "0")))
grid = device.compute_with_storage_grid_size()
G = ttnn.CoreGrid(y=grid.y, x=grid.x)
arch = str(device.arch()).lower()
CK = ttnn.BlackholeComputeKernelConfig if "blackhole" in arch and hasattr(ttnn, "BlackholeComputeKernelConfig") else ttnn.WormholeComputeKernelConfig


def ck(fidelity):
    return CK(math_fidelity=fidelity, math_approx_mode=False, fp32_dest_acc_en=False, packer_l1_acc=True)


def bits(t):
    return ttnn.to_torch(t).to(torch.bfloat16).reshape(-1).view(torch.int16)


def timed(fn, iters=20):
    for _ in range(3):
        fn()
    ttnn.synchronize_device(device)
    t0 = time.perf_counter()
    for _ in range(iters):
        fn()
    ttnn.synchronize_device(device)
    return (time.perf_counter() - t0) * 1e6 / iters


def resolved_fidelity(a, b, program_config=None):
    p = ttnn.MatmulParams()
    if program_config is not None:
        p.program_config = program_config
    c = ttnn.create_matmul_attributes(a, b, p, []).compute_kernel_config
    return getattr(c, "math_fidelity", c)


def own_2d_config(m, k, n):
    # The config ttnn picks for matmul(a, b) with nothing given, at a shape that takes the
    # all-DRAM 2D branch of create_simple_matmul_program_config: the full grid, per-core block
    # ceil(Mt / grid.y) x ceil(Nt / grid.x), in0_block_w shrunk to 1 so the block fits L1.
    # Its output is bit-identical to matmul(a, b) once the fidelity is set back to HiFi2 (see below).
    mt, nt = m // 32, n // 32
    pm, pn = -(-mt // grid.y), -(-nt // grid.x)
    return ttnn.MatmulMultiCoreReuseMultiCastProgramConfig(
        compute_with_storage_grid_size=(grid.x, grid.y), in0_block_w=1,
        out_subblock_h=pm, out_subblock_w=1, out_block_h=pm, out_block_w=pn,
        per_core_M=pm, per_core_N=pn, transpose_mcast=False, fused_activation=None)


print(f"arch {arch}, compute grid {grid.x}x{grid.y}")
g = torch.Generator().manual_seed(17)
for m, k, n in [(2048, 4096, 14336), (32, 4096, 14336)]:
    a = ttnn.from_torch((torch.randn([1, 1, m, k], generator=g) * k**-0.5).to(torch.bfloat16), layout=ttnn.TILE_LAYOUT, device=device)
    b = ttnn.from_torch((torch.randn([1, 1, k, n], generator=g) * k**-0.5).to(torch.bfloat16), layout=ttnn.TILE_LAYOUT, device=device)
    cfg = own_2d_config(m, k, n) if m > 32 else None
    print(f"\n== {m}x{k}x{n} bf16")
    print(f"fidelity create_matmul_attributes resolves with nothing given : {resolved_fidelity(a, b)}")
    if cfg is not None:
        print(f"fidelity it resolves with a program config given             : {resolved_fidelity(a, b, cfg)}")
        print(f"program config used below: {cfg}")
    ref = bits(ttnn.matmul(a, b))
    # (label, fn, reference for the second column: the same call form at an explicit HiFi2)
    hifi2 = ck(ttnn.MathFidelity.HiFi2)
    rows = [("matmul(a, b)", lambda: ttnn.matmul(a, b), None)]
    if cfg is not None:
        cfg_h2 = bits(ttnn.matmul(a, b, program_config=cfg, compute_kernel_config=hifi2))
        rows.append(("matmul(a, b, program_config=cfg)", lambda: ttnn.matmul(a, b, program_config=cfg), cfg_h2))
        for name, fid in [("HiFi2", ttnn.MathFidelity.HiFi2), ("LoFi", ttnn.MathFidelity.LoFi), ("HiFi4", ttnn.MathFidelity.HiFi4)]:
            rows.append((f"  same, compute_kernel_config={name}", (lambda f=fid: ttnn.matmul(a, b, program_config=cfg, compute_kernel_config=ck(f))), cfg_h2))
    grid_h2 = bits(ttnn.matmul(a, b, core_grid=G, compute_kernel_config=hifi2))
    rows.append(("matmul(a, b, core_grid=G)", lambda: ttnn.matmul(a, b, core_grid=G), grid_h2))
    rows.append(("  same, compute_kernel_config=HiFi2", lambda: ttnn.matmul(a, b, core_grid=G, compute_kernel_config=hifi2), grid_h2))
    print(f"{'form':44s} {'us':>8s}   {'words != matmul(a, b)':>22s}   {'words != same form at HiFi2':>28s}   (of {ref.numel()})")
    for name, fn, h2 in rows:
        out = bits(fn())
        d_ref = int((out != ref).sum())
        d_h2 = "" if h2 is None else f"{int((out != h2).sum()):>28d}"
        print(f"{name:44s} {timed(fn):8.1f}   {d_ref:>22d}   {d_h2:>28s}")
ttnn.close_device(device)
```

</details>

## Not established

- How many of the 146 shipped calls have bf16 inputs; bfp8 inputs are LoFi in every form.
- `linear` with a bias, `addmm` and `matmul_batched_weights` go through the same attribute builder (`matmul.cpp:253`, `:307`, `:502`) and were not run.
- Whether the models that always pass program configs (tt_transformers and the like) intend LoFi; for them this is the status quo.
- `sparse_matmul` builds a `MatmulParams` and calls the same builder (`sparse_matmul_device_operation.cpp:392`); it always carries a program config, so it is LoFi in every form; not measured.

Hardware: P300 (device 0), tt-metal main `b2e77906b92`; N300, tt-metal `bef164b30a9`.
